File: Examples/Carrier generator/carrier_generator.c
/*
Goal:
    This example is for educational purposes only, and is fully commented
    This example only generate a carrier on passed frequency or on default frequency 105.0 MHz
    Turn your car radio on and check if there is silence on that frequency

Usage:
    Compile:
        gcc carrier_generator.c -o carrier_generator -L /opt/vc/lib -l bcm_host -I /opt/vc/include
    Run:
        sudo ./carrier_generator
        sudo ./carrier_generator 106200000
    Stop:
        ctrl-\
        ctrl-c

References:
    https://github.com/raspberrypi/documentation/raw/master/hardware/raspberrypi/bcm2835/BCM2835-ARM-Peripherals.pdf
        Section 6 General Purpose I/O (GPIO)
        Pages:
            -  90 (Register View)
            -  92 (GPIO Alternate function select register 0)
            - 105 (General Purpose GPIO Clocks, MASH dividers)
            - 107 (Clock Manager General Purpose Clock Control)
            - 108 (Clock Manager General Purpose Clock Divisors)

    https://github.com/raspberrypi/userland/blob/master/host_applications/linux/libs/bcm_host/bcm_host.c
        Broadcom functions to get physical addresses to allow compatibility on RPi 1 and successors:
            - unsigned bcm_host_get_peripheral_address(void)
            - unsigned bcm_host_get_peripheral_size(void)
            - unsigned bcm_host_get_sdram_address(void)

    https://pinout.xyz/pinout/gpclk
    https://pinout.xyz/pinout/pin7_gpio4

License:
    GPL3

Author:
    Lorenzo Santina (BigNerd95)
*/

#include <bcm_host.h>           // Broadcom library to support any version of RaspBerry
#include <stdio.h>              // Input/Output functions
#include <stdint.h>             // Integer types having specified widths
#include <unistd.h>             // Function pause to wait for signals
#include <time.h>               // Function nanosleep
#include <fcntl.h>              // Function open
#include <sys/types.h>          // Sometimes needed by fcntl
#include <sys/stat.h>           // Sometimes needed by fcntl
#include <sys/mman.h>           // Function mmap



/***********************
* Peripheral constants *
***********************/

// All the following are BUS addresses

// Peripheral BUS base address
#define BUS_BASE_ADDRESS 0x7E000000


// GPIO constants

#define FLAGS_PER_REGISTER 10    // There are 10 pin flags per register

// General Purpose I/O function select registers
const unsigned GPFSEL[] = {
    0x7E200000,     // GPFSEL0 register, BCM pin  0 to  9 (we use 4 [GPIO4 pin 7])
    0x7E200004,     // GPFSEL1 register, BCM pin 10 to 19 (not used in this example)
    0x7E200008,     // GPFSEL2 register, BCM pin 20 to 29 (not used in this example)
    0x7E20000C,     // GPFSEL3 register, BCM pin 30 to 39 (not used in this example)
    0x7E200010,     // GPFSEL4 register, BCM pin 40 to 49 (not used in this example)
    0x7E200014      // GPFSEL5 register, BCM pin 50 to 53 (not used in this example)
};

// GPIO functions flags (each flag is 3 bit long)
typedef enum {
    GP_INPUT  = 0b000,
    GP_OUTPUT = 0b001,
    GP_AF0    = 0b100,
    GP_AF1    = 0b101,
    GP_AF2    = 0b110,
    GP_AF3    = 0b111,
    GP_AF4    = 0b011,
    GP_AF5    = 0b010
} GP_FUNCTION;


// CLOCK constants

// Clock Manager General Purpose Clocks Control registers
const unsigned CM_GPCTL[] = {
    0x7E101070,    // CM_GP0CTL (GPCLK0) Control register [Alternative Function 0 (GP_AF0) of GPIO4 (pin  7)]
    0x7E101078,    // CM_GP1CTL (GPCLK1) Control register [Alternative Function 0 (GP_AF0) of GPIO5 (pin 29)] (not used in this example)
    0x7E101080     // CM_GP2CTL (GPCLK2) Control register [Alternative Function 0 (GP_AF0) of GPIO6 (pin 31)] (not used in this example)
};

// Clock Manager General Purpose Clock Divisors registers
const unsigned CM_GPDIV[] = {
    0x7E101074,    // CM_GP0DIV (GPCLK0) Divisor register [Alternative Function 0 (GP_AF0) of GPIO4 (pin  7)]
    0x7E10107C,    // CM_GP1DIV (GPCLK1) Divisor register [Alternative Function 0 (GP_AF0) of GPIO5 (pin 29)] (not used in this example)
    0x7E101084     // CM_GP2DIV (GPCLK2) Divisor register [Alternative Function 0 (GP_AF0) of GPIO6 (pin 31)] (not used in this example)
};

// Clock Managers General Purpose Clocks
typedef enum {
    CM_GPCLK0 = 0,
    CM_GPCLK1 = 1,
    CM_GPCLK2 = 2
} CM_GPCLK;

// Clock Manager flags
#define CM_PASSWD (0x5A << 24)
#define CM_ENAB   (1 << 4)
#define CM_BUSY   (1 << 7)

// Clock Manager Control register mash flags
typedef enum {
    CM_MASH_INT = (0 << 9),
    CM_MASH_1S  = (1 << 9),
    CM_MASH_2S  = (2 << 9),
    CM_MASH_3S  = (3 << 9)
} CM_MASH;

// Clock Manager Control register source flags
typedef enum {
  CM_GND           = 0,
  CM_OSCILLATOR    = 1,
  CM_TESTDEBUD0    = 2,
  CM_TESTDEBUG1    = 3,
  CM_PLLA          = 4,
  CM_PLLC          = 5,
  CM_PLLD          = 6,
  CM_HDMI          = 7
} CM_SRC;

#define PLLD_FREQ     500000000.00 // PLLD clock source frequency (500MHz)


// Address conversion macro

void* VIRTUAL_BASE_ADDRESS = NULL;       // Global variable where peripherals virtual base address will be contained after mapping physical address
#define CONVERT(address)    ((unsigned) (address) - BUS_BASE_ADDRESS + VIRTUAL_BASE_ADDRESS)  // convert BUS address to virtual address
#define SET(address, value) ((*(uint32_t*) CONVERT(address)) = (value))          // set value to BUS address
#define GET(address)        (*(uint32_t*) CONVERT(address))                      // get value from BUS address



/***********************
* Peripheral functions *
***********************/

// Map a physical address in the virtual address space of this process
void* map_memory(unsigned address, size_t size) {

    int fd = open("/dev/mem", O_RDWR | O_SYNC);     // Open /dev/mem to map the physical address
    if (fd < 0){                                    // Check if open is failed
        puts("Can't open /dev/mem\nRun as root!");
        exit(-1);
    }

    void* vaddr = mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_SHARED, fd, address);  // map physical address from /dev/mem fd with given size
    if (vaddr == MAP_FAILED){                       // Check if mmap is failed
        puts("Failed to map memory");
        exit(-1);
    }

    close(fd);                                      // Close /dev/mem fd
    return vaddr;                                   // Return mapped address in this process address space
}

// Maps the correct peripheral physical address in the virtual address space of this process
void map_peripheral(){
    unsigned peripheral_address = bcm_host_get_peripheral_address();  // Broadcom function to get peripheral physical address (so it support any Raspberry Pi version)
    unsigned peripheral_size = bcm_host_get_peripheral_size();        // Broadcom function to get peripheral size
    VIRTUAL_BASE_ADDRESS = map_memory(peripheral_address, peripheral_size);    // Map peripheral memory
}

// Map peripherals if they are not mapped
void check_peripheral(){
    if (!VIRTUAL_BASE_ADDRESS)
        map_peripheral();
}

// Set the value of the given BUS addres
void reg_set(unsigned address, uint32_t value){
    check_peripheral();             // Check if peripherals are mapped
    SET(address, value);            // Set the value
}

// Get the value of the given BUS address
uint32_t reg_get(unsigned address){
    check_peripheral();             // Check if peripherals are mapped
    return GET(address);            // Get the value
}

// Set gpio pin function (pin number is in BCM format)
void set_gp_func(unsigned int pin, GP_FUNCTION function){
    if (pin >= 0 && pin <= 53){
        unsigned int reg_number = pin / FLAGS_PER_REGISTER;     // Calculate register number
        unsigned int pin_number = pin % FLAGS_PER_REGISTER;     // Calculate pin number of this register

        uint32_t value = reg_get(GPFSEL[reg_number]);           // Get GPFSELn register value
        value &= ~(   0b111 << 3 * pin_number);                 // Zeroes out the 3 bits relative to this pin
        value |=  (function << 3 * pin_number);                 // Set pin function
        reg_set(GPFSEL[reg_number], value);                     // Set GPFSELn register new value
    } else {
        puts("Invalid pin number!");
        exit(-1);
    }
}

// Disable clock generator and wait until busy flag turns off
void stop_clock(CM_GPCLK gpclk_number, CM_SRC clk_source){
    while (reg_get(CM_GPCTL[gpclk_number]) & CM_BUSY)              // Unless busy flag turns off
        reg_set(CM_GPCTL[gpclk_number], CM_PASSWD | clk_source);   // Disable clock generator
}

// Set frequency divisor, source, mash and enable clock generator
void start_clock(CM_GPCLK gpclk_number, CM_SRC clk_source, CM_MASH mash_stage){
    stop_clock(gpclk_number, clk_source);                                    // Disable clock generator before any changes
    reg_set(CM_GPCTL[gpclk_number], CM_PASSWD | mash_stage | clk_source);    // Set source and MASH

    while(!(reg_get(CM_GPCTL[gpclk_number]) & CM_BUSY))                                          // Unless busy flag turns on
        reg_set(CM_GPCTL[gpclk_number], CM_PASSWD | reg_get(CM_GPCTL[gpclk_number]) | CM_ENAB);  // Enable clock generator
}

// Set clock generator frequency
void set_clock_frequency(CM_GPCLK gpclk_number, uint32_t frequency){
    uint32_t clock_divisor = ((float)(PLLD_FREQ / frequency)) * (1 << 12);   // Calculate clock frequency divisor (both integer and fractional part)
    reg_set(CM_GPDIV[gpclk_number], CM_PASSWD | clock_divisor);              // Set frequency divisor
}



/***************
* Main program *
***************/

// Stop clock generator and reset GPIO4 to OUTPUT function
void exit_handler(int signum){
    puts("\nCleaning resources...");
    set_gp_func(4, GP_OUTPUT);        // Set GPIO4 (pin 7) as output
    stop_clock(CM_GPCLK0, CM_PLLD);   // Disable clock generator
    exit(signum);                     // Close process
}

// Set signal handler
void set_signal_handler(int signum, void (*handler)(int)){
    if (signal(signum, handler) == SIG_ERR)         // Set signal handler and check success
        printf("Set signal (%d) error", signum);
}

// Set handler to clean resources on exit signals
void run_forever(){
    set_signal_handler(SIGQUIT, exit_handler);      // Set SIGQUIT (ctrl-\) signal handler to reset gpio on exit
    set_signal_handler(SIGINT,  exit_handler);      // Set SIGINT  (ctrl-c) signal handler to reset gpio on exit
    pause();                                        // Wait for signals (unistd function), blocked waiting (so the CPU is not overloaded)
}

int main(int argc, char **argv){

    uint32_t carrier_frequency = 105000000;             // Default frequency 105.0 MHz
    if (argc > 1)
        carrier_frequency = atoi(argv[1]);              // Parse first argument as frequency if present

    set_gp_func(4, GP_AF0);                             // Set GPIO4 (pin 7) as clock (Alternative Function 0)
    start_clock(CM_GPCLK0, CM_PLLD, CM_MASH_1S);        // Enable GPCLK0 clock generator (because we are using GPIO4 [pin 7]) using PLLD clock source (500MHz) and 1-stage MASH (to support fractional divisor)
    set_clock_frequency(CM_GPCLK0, carrier_frequency);  // Set the clock frequency of GPCLK0
    printf("Transmitting carrier on %d Hz\n", carrier_frequency);

    run_forever();                                      // Keep this process alive
    return 0;
}

File: Examples/Carrier generator/carrier_generator.py
#!/usr/bin/env python3

import struct, mmap
import sys, signal


class Peripheral():

    # Peripheral BUS base address
    BUS_BASE_ADDRESS = 0x7E000000


    # GPIO

    # There are 10 pin flags per register
    FLAGS_PER_REGISTER = 10

    # General Purpose I/O function select registers
    GPFSEL = [
        0x7E200000, # GPFSEL0 BCM pin  0 to  9 (we use 4 [GPIO4 pin 7])
        0x7E200004, # GPFSEL1 BCM pin 10 to 19 (not used in this example)
        0x7E200008, # GPFSEL2 BCM pin 20 to 29 (not used in this example)
        0x7E20000C, # GPFSEL3 BCM pin 30 to 39 (not used in this example)
        0x7E200010, # GPFSEL4 BCM pin 40 to 49 (not used in this example)
        0x7E200014  # GPFSEL5 BCM pin 50 to 53 (not used in this example)
    ]

    # GPIO functions flags (each flag is 3 bit long)
    GP_INPUT  = 0b000
    GP_OUTPUT = 0b001
    GP_AF0    = 0b100
    GP_AF1    = 0b101
    GP_AF2    = 0b110
    GP_AF3    = 0b111
    GP_AF4    = 0b011
    GP_AF5    = 0b010


    # CLOCK
    
    # Clock Manager General Purpose Clocks Control registers
    CM_GPCTL = [
        0x7E101070, # CM_GP0CTL (GPCLK0) Control register [Alternative Function 0 (GP_AF0) of GPIO4 (pin  7)]
        0x7E101078, # CM_GP1CTL (GPCLK1) Control register [Alternative Function 0 (GP_AF0) of GPIO5 (pin 29)] (not used in this example)
        0x7E101080  # CM_GP2CTL (GPCLK2) Control register [Alternative Function 0 (GP_AF0) of GPIO6 (pin 31)] (not used in this example)
    ]

    # Clock Manager General Purpose Clock Divisors registers
    CM_GPDIV = [
        0x7E101074, # CM_GP0DIV (GPCLK0) Divisor register [Alternative Function 0 (GP_AF0) of GPIO4 (pin  7)]
        0x7E10107C, # CM_GP1DIV (GPCLK1) Divisor register [Alternative Function 0 (GP_AF0) of GPIO5 (pin 29)] (not used in this example)
        0x7E101084  # CM_GP2DIV (GPCLK2) Divisor register [Alternative Function 0 (GP_AF0) of GPIO6 (pin 31)] (not used in this example)
    ]

    # Clock Managers General Purpose Clocks
    CM_GPCLK0 = 0
    CM_GPCLK1 = 1
    CM_GPCLK2 = 2

    # Clock Manager flags
    CM_PASSWD = (0x5A << 24)
    CM_ENAB   = (1 << 4)
    CM_BUSY   = (1 << 7)

    # Clock Manager Control register mash flags
    CM_MASH_INT = (0 << 9)
    CM_MASH_1S  = (1 << 9)
    CM_MASH_2S  = (2 << 9)
    CM_MASH_3S  = (3 << 9)

    # Clock Manager Control register SOURCE flags
    CM_GND        = 0
    CM_OSCILLATOR = 1
    CM_TESTDEBUD0 = 2
    CM_TESTDEBUG1 = 3
    CM_PLLA       = 4
    CM_PLLC       = 5
    CM_PLLD       = 6
    CM_HDMI       = 7

    # PLLD clock source frequency (500MHz)
    PLLD_FREQ = 500000000


    def __init__(self):
        peripheral_address = Peripheral.__bcm_host_get_peripheral_address__()
        peripheral_size    = Peripheral.__bcm_host_get_peripheral_size__()
        self.PERIPHERAL_MEMORY = Peripheral.__map_peripheral_memory__(peripheral_address, peripheral_size)

    def __get_dt_ranges__(filename, offset):
        address = 0xFFFFFFFF
        with open(filename, "rb") as fp:
            fp.seek(offset, 0)
            buf = fp.read(4)
            if len(buf) == 4:
                address = struct.unpack('>I', buf)[0]
        return address

    def __bcm_host_get_peripheral_address__():
        address = Peripheral.__get_dt_ranges__("/proc/device-tree/soc/ranges", 4)
        if address == 0xFFFFFFFF:
            address = 0x20000000
        return address


    def __bcm_host_get_peripheral_size__():
        address = Peripheral.__get_dt_ranges__("/proc/device-tree/soc/ranges", 8)
        if address == 0xFFFFFFFF:
            address = 0x01000000
        return address

    def __map_peripheral_memory__(address, size):
        with open("/dev/mem", "r+b") as fp:
            return mmap.mmap(fp.fileno(), size, offset=address)

    # Convert BUS address to virtual address
    def __convert_address__(address):
        return address - Peripheral.BUS_BASE_ADDRESS

    # Set the value of the given BUS addres
    def __reg_set__(self, address, value):
        relative_address = Peripheral.__convert_address__(address)
        self.PERIPHERAL_MEMORY[relative_address : relative_address + 4] = struct.pack("<I", value)

    # Get the value of the given BUS address
    def __reg_get__(self, address):
        relative_address = Peripheral.__convert_address__(address)
        return struct.unpack("<I", self.PERIPHERAL_MEMORY[relative_address : relative_address + 4])[0]

    def print_reg(self, reg):
        print(hex(self.__reg_get__(reg)))

    # Set gpio pin function (pin number is in BCM format)
    def set_gp_func(self, pin, function):
        if pin in range(0, 54):
            reg_number = int(pin / Peripheral.FLAGS_PER_REGISTER)   # Calculate register number
            pin_number =     pin % Peripheral.FLAGS_PER_REGISTER    # Calculate pin number of this register

            value = self.__reg_get__(Peripheral.GPFSEL[reg_number]) # Get GPFSELn register value
            value &= ~(   0b111 << 3 * pin_number)                  # Zeroes out the 3 bits relative to this pin
            value |=  (function << 3 * pin_number)                  # Set pin function
            self.__reg_set__(Peripheral.GPFSEL[reg_number], value)  # Set GPFSELn register new value
        else:
            print("Invalid pin number!")

    def set_clock_frequency(self, gpclk_number, frequency):
        clock_divisor = (Peripheral.PLLD_FREQ / frequency) * (1 << 12)   # Calculate clock frequency divisor (both integer and fractional part)
        self.__reg_set__(Peripheral.CM_GPDIV[gpclk_number], Peripheral.CM_PASSWD | int(clock_divisor))      # Set frequency divisor

    # Set frequency divisor, source, mash and enable clock generator
    def start_clock(self, gpclk_number, clk_source, mash_stage):
        self.stop_clock(gpclk_number, clk_source)
        self.__reg_set__(Peripheral.CM_GPCTL[gpclk_number], Peripheral.CM_PASSWD | mash_stage | clk_source) # Set source and MASH
        
        while not self.__reg_get__(Peripheral.CM_GPCTL[gpclk_number]) & Peripheral.CM_BUSY:                 # Unless busy flag turns on
            self.__reg_set__(Peripheral.CM_GPCTL[gpclk_number], Peripheral.CM_PASSWD | self.__reg_get__(Peripheral.CM_GPCTL[gpclk_number]) | Peripheral.CM_ENAB) # Enable clock generator
    
    # Disable clock generator and wait until busy flag turns off
    def stop_clock(self, gpclk_number, clk_source):
        while self.__reg_get__(Peripheral.CM_GPCTL[gpclk_number]) & Peripheral.CM_BUSY:            # Unless busy flag turns off
            self.__reg_set__(Peripheral.CM_GPCTL[gpclk_number], Peripheral.CM_PASSWD | clk_source) # Disable clock generator




# Stop clock generator and reset GPIO4 to OUTPUT function
def exit_handler(signum, frame):
    print("Cleaning resources...")
    per.set_gp_func(4, Peripheral.GP_OUTPUT)                 # Set GPIO4 (pin 7) as output
    per.stop_clock(Peripheral.CM_GPCLK0, Peripheral.CM_PLLD) # Disable clock generator
    sys.exit(signum)                                         # Close process

# Set handler to clean resources on exit signals
def run_forever():
    signal.signal(signal.SIGQUIT, exit_handler)      # Set SIGQUIT (ctrl-\) signal handler to reset gpio on exit
    signal.signal(signal.SIGINT,  exit_handler)      # Set SIGINT  (ctrl-c) signal handler to reset gpio on exit
    signal.pause()                                   # Wait for signals (blocked waiting, so the CPU is not overloaded)


if __name__ == "__main__":

    carrier_frequency = 105000000
    if len(sys.argv) > 1:
        carrier_frequency = int(sys.argv[1])

    per = Peripheral()
    per.set_gp_func(4, Peripheral.GP_AF0)
    per.start_clock(Peripheral.CM_GPCLK0, Peripheral.CM_PLLD, Peripheral.CM_MASH_1S)
    per.set_clock_frequency(Peripheral.CM_GPCLK0, carrier_frequency)

    print("Transmitting carrier on " + str(carrier_frequency) + " Hz")

    run_forever()
